File: python/code_039/opencv_039.py
import cv2 as cv
import numpy as np


def template_demo():
    src = cv.imread("./test.png")
    tpl = cv.imread("./test01.png")
    cv.imshow("input", src)
    cv.imshow("tpl", tpl)
    th, tw = tpl.shape[:2]
    result = cv.matchTemplate(src, tpl, cv.TM_CCORR_NORMED)
    cv.imshow("result", result)
    cv.imwrite("D:/039_003.png", np.uint8(result*255))
    t = 0.98
    loc = np.where(result > t)

    for pt in zip(*loc[::-1]):
        cv.rectangle(src, pt, (pt[0] + tw, pt[1] + th), (255, 0, 0), 1, 8, 0)
    cv.imshow("llk-demo", src)
    cv.imwrite("D:/039_004.png", src)


template_demo()
cv.waitKey(0)
cv.destroyAllWindows()

File: python/code_039/opencv_039.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;
const float t = 0.95;
int main(int artc, char** argv) {
	Mat src = imread("./test01.png");
	Mat tpl = imread("./test.png");
	if (src.empty() || tpl.empty()) {
		printf("could not load image...\n");
		return -1;
	}
	imshow("input", src);
	imshow("tpl", tpl);

	int result_h = src.rows - tpl.rows + 1;
	int result_w = src.cols - tpl.cols + 1;
	Mat result = Mat::zeros(Size(result_w, result_h), CV_32FC1);
	matchTemplate(src, tpl, result, TM_CCOEFF_NORMED);
	imshow("result image", result);
	int h = result.rows;
	int w = result.cols;
	for (int row = 0; row < h; row++) {
		for (int col = 0; col < w; col++) {
			float v = result.at<float>(row, col);
			// printf("v = %.2f\n", v);
			if (v > t) {
				rectangle(src, Point(col, row), Point(col + tpl.cols, row + tpl.rows), Scalar(255, 0, 0), 1, 8, 0);
			}
		}
	}
	imshow("template matched result", src);

	waitKey(0);
	return 0;
}
